broker/src/main/java/io/zeebe/broker/system/partitions/impl/StateControllerImpl.java

```diff
@@ -71,29 +71,36 @@ public StateControllerImpl(
   @Override
   public Optional<TransientSnapshot> takeTransientSnapshot(final long lowerBoundSnapshotPosition) {
     if (!isDbOpened()) {
+      LOG.warn(
+          "Expected to take snapshot for lower bound position {}, but database was closed.",
+          lowerBoundSnapshotPosition);
       return Optional.empty();
     }
 
-    final long exportedPosition = exporterPositionSupplier.applyAsLong(openDb());
-    final long snapshotPosition = Math.min(exportedPosition, lowerBoundSnapshotPosition);
-
+    final long snapshotPosition = determineSnapshotPosition(lowerBoundSnapshotPosition);
     final var optionalIndexed = entrySupplier.getIndexedEntry(snapshotPosition);
+    if (optionalIndexed.isEmpty()) {
+      LOG.warn(
+          "Expected to find an indexed entry for given snapshot position {}, but found no matching indexed entry which contains this position.",
+          lowerBoundSnapshotPosition);
+      return Optional.empty();
+    }
 
-    final Long previousSnapshotIndex =
+    final var snapshotIndexedEntry = optionalIndexed.get();
+    final long previousSnapshotIndex =
         store.getLatestSnapshot().map(PersistedSnapshot::getCompactionBound).orElse(-1L);
+    if (snapshotIndexedEntry.index() == previousSnapshotIndex) {
+      LOG.debug(
+          "Previous snapshot was taken for the same indexed entry {}, will not take snapshot.",
+          snapshotIndexedEntry);
+      return Optional.empty();
+    }
 
-    final var optTransientSnapshot =
-        optionalIndexed
-            .filter(indexed -> indexed.index() != previousSnapshotIndex)
-            .map(
-                indexed ->
-                    store.newTransientSnapshot(
-                        indexed.index(),
-                        indexed.entry().term(),
-                        WallClockTimestamp.from(System.currentTimeMillis())));
-
-    optTransientSnapshot.ifPresent(this::createSnapshot);
```

Comment:
I think this was removed by mistake :slightly_smiling_face: . This is probably why tests are failing.

Reply:
Ups :D yeah thanks will revert that :see_no_evil: